streams/shared/src/main/scala/scalaz/stream/ZStream.scala

```diff
@@ -170,14 +170,98 @@ trait ZStream[-R, +E, +A] extends Serializable { self =>
+              //     - If one of them failed, all other fibers are interrupted
+              //     - If they all succeeded, Take.End is enqueued
+              //   - On error, the driver stream interrupts all inner fibers and emits a
+              //     Take.Fail value
+              //   - On interruption, the driver stream interrupts all inner fibers and
+              //     shuts down the queue
+              // - Inner fibers enqueue Take values from their streams to the output queue
+              //   - On error, an inner fiber enqueues a Take.Error value and signals its failure
+              //     with a promise. The driver will pick that up and interrupt all other fibers.
+              //   - On interruption, an inner fiber does nothing
+              //   - On completion, an inner fiber does nothing
+              _ <- Managed.make {
+                    Promise.make[E1, Nothing].flatMap { innerFailure =>
+                      self.foreach { a =>
+                        permits.withPermit {
+                          f(a)
+                            .foreach(b => out.offer(Take.Value(b)).unit)
+                            .onError { cause =>
+                              cause.failureOrCause.fold(
+                                e => out.offer(Take.Fail(e)) *> innerFailure.fail(e),
+                                _ => ZIO.unit
+                              )
+                            }
+                            .fork
+                            .unit
+                        }
+                      }.foldCauseM(
+                          cause =>
+                            cause.failureOrCause.fold(
+                              // The driver stream failed.
+                              e => out.offer(Take.Fail(e)),
+                              // The driver stream had a defect.
+                              c => out.shutdown *> ZIO.halt(c)
+                            ),
+                          _ =>
+                            for {
+                              innerFibers <- ZIO.children
+                              _ <- innerFailure.await.raceWith(Fiber.joinAll(innerFibers))(
+                                    // One of the inner fibers failed. It already enqueued its failure,
+                                    // so we don't need to do anything except interrupt all the other fibers.
+                                    leftDone = (_, _) => Fiber.interruptAll(innerFibers),
+                                    // All fibers completed successfully, so we just need to signal that
+                                    // we're done.
+                                    rightDone = (_, failureAwait) => out.offer(Take.End) *> failureAwait.interrupt
+                                  )
+                            } yield ()
+                        )
+                        .onInterrupt(ZIO.children.flatMap(Fiber.interruptAll) *> out.shutdown)
```

Comment:
Isn't this redundant with the line below?

Reply:
`supervise` is the one that ends a supervisor that interrupts all children; `supervised` just enables supervision children are tracked and introspectable by `ZIO.children` :-)

Reply:
😰 ah yeah, I remember. I have some catching up to do!

Reply:
We probably need better names too... like `enableSupervision` and `cleanupChildren` or something...

Reply:
yeah. I am constantly confused between the two ;-)